api/src/db/model/namespace.ts

```diff
@@ -230,72 +235,100 @@ export default class NamespaceModel extends Model {
+      logger.error(`${message}, err = ${err.message}`);
 
-      const clusterNamespaces: (ClusterNamespace | undefined)[] = clResults.map(
-        (cl) => cl.pop()
-      );
+      throw err;
+    }
+  }
 
-      const quotaSizes: NameSpacesQuotaSize = {
-        quotaCpuSize: [],
-        quotaMemorySize: [],
-        quotaStorageSize: [],
-        quotaSnapshotSize: [],
+  async getProjectSetQuotaSize(
+    profileId: number,
+    clusterId: number
+  ): Promise<any> {
+    const query = {
+      text: `
+      SELECT 
+        name as namespaceName, 
+        json_build_object(
+          'quotaCpuSize', quota_cpu_size, 
+          'quotaMemorySize', quota_memory_size, 
+          'quotaStorageSize', quota_storage_size, 
+          'quotaSnapshotSize', quota_snapshot_size ) as quota_info
+    FROM cluster_namespace
+    JOIN namespace ON cluster_namespace.namespace_id = namespace.id 
+    WHERE namespace.profile_id = $1 AND cluster_namespace.cluster_id = $2;
+    `,
+      values: [],
+    };
+    try {
+      const ClustersnamespaceQuotaSize = await this.runQuery({
+        ...query,
+        values: [profileId, clusterId],
+      });
+
+      const projectQuotaSize: ProjectQuotaSize = {
+        dev: {
+          quotaCpuSize: QuotaSize.Small,
+          quotaMemorySize: QuotaSize.Small,
+          quotaStorageSize: QuotaSize.Small,
+          quotaSnapshotSize: QuotaSize.Small,
+        },
+        test: {
+          quotaCpuSize: QuotaSize.Small,
+          quotaMemorySize: QuotaSize.Small,
+          quotaStorageSize: QuotaSize.Small,
+          quotaSnapshotSize: QuotaSize.Small,
+        },
+        tools: {
+          quotaCpuSize: QuotaSize.Small,
+          quotaMemorySize: QuotaSize.Small,
+          quotaStorageSize: QuotaSize.Small,
+          quotaSnapshotSize: QuotaSize.Small,
+        },
+        prod: {
+          quotaCpuSize: QuotaSize.Small,
+          quotaMemorySize: QuotaSize.Small,
+          quotaStorageSize: QuotaSize.Small,
+          quotaSnapshotSize: QuotaSize.Small,
+        },
       };
-
-      clusterNamespaces.forEach(
-        (clusterNamespace: ClusterNamespace | undefined): void => {
-          if (!clusterNamespace) {
-            return;
-          }
-          const {
-            quotaCpuSize,
-            quotaMemorySize,
-            quotaStorageSize,
-            quotaSnapshotSize,
-          } = clusterNamespace;
-
-          quotaSizes.quotaCpuSize.push(quotaCpuSize);
-          quotaSizes.quotaMemorySize.push(quotaMemorySize);
-          quotaSizes.quotaStorageSize.push(quotaStorageSize);
-          quotaSizes.quotaSnapshotSize.push(quotaSnapshotSize);
+      // TODO cluster duplication only show one
+      ClustersnamespaceQuotaSize.forEach((namespaceQuotaInfo) => {
+        const namespace = getLicenseplatPostfix(
+          namespaceQuotaInfo.namespacename
+        );
+
+        try {
+          projectQuotaSize[namespace] = namespaceQuotaInfo.quotaInfo;
```

Comment:
this mutable javascript hurts my soul a bit. Anyway to do this immutably?

Reply:
yeah, healing your soul is always my priority.